db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,131 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client> client, sstri
+    auto s = std::make_unique<optimized_optional<abort_source::subscription>>(_as.subscribe([&]() noexcept {
+        auto h = as_gate.hold();
+        std::ignore = sharded_aborter.invoke_on_all([ex = _as.abort_requested_exception_ptr()](abort_source& aborter) {
+            aborter.request_abort_ex(ex);
+        }).finally([h = std::move(h)] {});
+    }));
+
+    // This will capture any exception from the backup process.
+    std::exception_ptr ret_ex;
+    co_await sharded_aborter
+        .invoke_on_all([this, &backup_files, &counter, &sharded_aborter](abort_source& aborter) -> future<> {
+            auto shard_id = this_shard_id();
+            gate uploads;
+            std::exception_ptr ex;
+            try {
+                auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+
+                // Process backup files continuously until:
+                //   1. There are no more files.
+                //   2. An exception is thrown.
+                //   3. An abort is requested.
+                while (true) {
+                    // One of two things happened: the `upload_component` threw an exception, `aborter`'s exception was set.
+                    // The later can happen if the `aborter` was requested to abort the backup by task's abort source or another shard requested to abort since
+                    // it got an exception
+                    if (ex || aborter.abort_requested()) {
+                        break;
+                    }
+
+                    // Fetching files sequentially within each shard allows less loaded shards to take on more work, compensating for those that are lagging behind.
+                    auto name_idx = counter++;
+                    if (name_idx >= backup_files.size()) {
+                        break;
+                    }
+                    auto gh = uploads.hold();
+                    // Pre-upload break point. For testing abort in actual s3 client usage.
+                    co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+                    // Start uploading the component asynchronously.
+                    // Future is ignored here because:
+                    // - The client (cln) is maintained via shared ownership.
+                    // - Filenames are passed by value.
+                    // - The abort source remains valid for the duration of the function.
+                    // - The gate ensures that execution remains until completion.
+                    std::ignore = upload_component(cln, aborter, _progress_per_shard[shard_id], std::get<1>(backup_files[name_idx]))
+                                      .handle_exception([&ex](std::exception_ptr e) {
+                                          // Record the exception to eventually abort the rest of the backup.
```

Comment:
This "eventually" is too way ahead in time. If there's a bunch of bg uploads spawned, then the main fiber sits in co_await uploads.close() below, and in case any upload fails, the remaining won't be aborted until all the others finish

Reply:
Invoking abortion on shards here too